delivery_carrier_multi_zip/models/delivery_carrier.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2022 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    zip_option = fields.Selection(
+        selection=[
+            ("prefix", "Prefixes"),
+            ("range", "Ranges"),
+        ],
+        default="prefix",
+    )
+    zip_range_ids = fields.One2many(
+        comodel_name="delivery.carrier.zip",
+        inverse_name="carrier_id",
+        string="ZIP codes",
+    )
+
+    def _match_address(self, partner):
+        """Match as well by zip intervals if they are present."""
+        res = super()._match_address(partner)  # it has self.ensure_one()
+        if res and self.zip_range_ids:
```

Comment:
Do this only if the option is selected, not depending on the entries (as they may exist, but the option has been changed).